scipy/optimize/minpack.py

```diff
@@ -510,28 +523,34 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
         values will all be 1 (if the number of parameters for the function
         can be determined using introspection, otherwise a ValueError
         is raised).
-    sigma : None or M-length sequence, optional
-        If not None, the uncertainties in the ydata array. These are used as
-        weights in the least-squares problem
-        i.e. minimising ``np.sum( ((f(xdata, *popt) - ydata) / sigma)**2 )``
-        If None, the uncertainties are assumed to be 1.
+    sigma : None or M-length sequence or MxM array, optional
+        Determines the uncertainty in `ydata`.
+        If we define measurement errors as
+        ``e = ydata - f(xdata, *popt)``, then
+
+            - If `sigma` is 1-d, it determines standard deviations of errors `e`
+              and the optimized function is ``sum((e / sigma) ** 2)``.
+            - If `sigma` is 2-d, it determines the covariance matrix of errors `e`
+              and the optimized function is ``e.T.dot(inv(sigma)).dot(e)``.
+
+        None (default) is equivalent of 1-d `sigma` filled with ones.
     absolute_sigma : bool, optional
-        If False, `sigma` denotes relative weights of the data points.
-        The returned covariance matrix `pcov` is based on *estimated*
-        errors in the data, and is not affected by the overall
-        magnitude of the values in `sigma`. Only the relative
-        magnitudes of the `sigma` values matter.
-
-        If True, `sigma` describes one standard deviation errors of
-        the input data points. The estimated covariance in `pcov` is
-        based on these values.
+        If False, `sigma` is used in a relative sense. The returned parameter
```

Comment:
About 2, I do not think that there are two scalings. The optimization is solved by using the errors as they are given by the user. It is only at the end that the routine checks whether absolute_sigma is False. If it is, then while determining what pcov to return, the errors are scaled so that the reduced chi-squared is unity. At least that is my understanding of the code as written currently.

3) I like to mention about the degrees of freedom, and the subtraction of parameters explicitly to make it super clear why this is done. The idea of obtaining reduced chi-squared close to unity is typically well known.

https://en.wikipedia.org/wiki/Goodness_of_fit